src/librustc_typeck/astconv.rs

```diff
@@ -2395,33 +2398,61 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         ty
     }
 
-    pub fn ty_of_arg(&self,
-                     ty: &hir::Ty,
-                     expected_ty: Option<Ty<'tcx>>)
-                     -> Ty<'tcx>
-    {
-        match ty.kind {
-            hir::TyKind::Infer if expected_ty.is_some() => {
-                self.record_ty(ty.hir_id, expected_ty.unwrap(), ty.span);
-                expected_ty.unwrap()
-            }
-            _ => self.ast_ty_to_ty(ty),
+    pub fn ty_of_arg(
+        &self,
+        ty: &hir::Ty,
+        expected_ty: Option<Ty<'tcx>>,
+    ) -> Ty<'tcx> {
+        if crate::collect::is_infer_ty(ty) && expected_ty.is_some() {
+            self.record_ty(ty.hir_id, expected_ty.unwrap(), ty.span);
+            expected_ty.unwrap()
+        } else {
+            self.ast_ty_to_ty(ty)
         }
     }
 
-    pub fn ty_of_fn(&self,
-                    unsafety: hir::Unsafety,
-                    abi: abi::Abi,
-                    decl: &hir::FnDecl)
-                    -> ty::PolyFnSig<'tcx> {
+    pub fn ty_of_fn(
+        &self,
+        unsafety: hir::Unsafety,
+        abi: abi::Abi,
+        decl: &hir::FnDecl,
+        generic_params: &[hir::GenericParam],
+        ident_span: Option<Span>,
+    ) -> ty::PolyFnSig<'tcx> {
         debug!("ty_of_fn");
 
         let tcx = self.tcx();
-        let input_tys =
-            decl.inputs.iter().map(|a| self.ty_of_arg(a, None));
 
+        // We proactively collect all the infered type params to emit a single error per fn def.
+        let mut placeholder_types = vec![];
+        let mut output_placeholder_types = vec![];
+
+        let input_tys = decl.inputs.iter().map(|a| {
+            let mut visitor = PlaceholderHirTyCollector::new();
+            visitor.visit_ty(&a);
+            if visitor.0.is_empty() || self.allow_ty_infer() {
+                self.ty_of_arg(a, None)
+            } else {
+                placeholder_types.extend(visitor.0);
+                tcx.types.err
+            }
+        });
         let output_ty = match decl.output {
-            hir::Return(ref output) => self.ast_ty_to_ty(output),
+            hir::Return(ref output) => {
+                let mut visitor = PlaceholderHirTyCollector::new();
+                visitor.visit_ty(output);
+                let is_infer = if let hir::TyKind::Infer = output.kind {
+                    true
+                } else {
+                    false
+                };
+                if (is_infer || !visitor.0.is_empty()) && !self.allow_ty_infer() {
+                    output_placeholder_types.extend(visitor.0);
+                    tcx.types.err
+                } else {
+                    self.ast_ty_to_ty(output)
+                }
+            }
```

Comment:
I want to clean this up, but when I just did I broke the logic in a weird way. Need to dig deeper.

Reply:
@Centril I think with the latest changes this is way more reasonable.